core/src/main/scala/kafka/raft/KafkaMetadataLog.scala

```diff
@@ -147,18 +199,106 @@ class KafkaMetadataLog(
   }
 
   override def createSnapshot(snapshotId: raft.OffsetAndEpoch): RawSnapshotWriter = {
-    FileRawSnapshotWriter.create(log.dir.toPath, snapshotId)
+    // TODO: Talk to Jason about truncation past the high-watermark since it can lead to truncation past snapshots.
+    // This can result in the leader having a snapshot that is less that the follower's snapshot. I think that the Raft
+    // Client checks against this and aborts. If so, then this check and exception is okay.
+
+    // Do let the state machine create snapshots older than the latest snapshot
+    latestSnapshotId().ifPresent { latest =>
+      if (latest.epoch > snapshotId.epoch || latest.offset > snapshotId.offset) {
+        // Since snapshots are less than the high-watermark absolute offset comparison is okay.
+        throw new IllegalArgumentException(
+          s"Attemting to create a snapshot ($snapshotId) that is not greater than the latest snapshot ($latest)"
+        )
+      }
+    }
+
+    FileRawSnapshotWriter.create(log.dir.toPath, snapshotId, Optional.of(this))
   }
 
   override def readSnapshot(snapshotId: raft.OffsetAndEpoch): Optional[RawSnapshotReader] = {
     try {
-      Optional.of(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
+      if (snapshotIds.contains(snapshotId)) {
+        Optional.of(FileRawSnapshotReader.open(log.dir.toPath, snapshotId))
+      } else {
+        Optional.empty()
+      }
     } catch {
-      case e: NoSuchFileException => Optional.empty()
+      case e: NoSuchFileException =>
+        Optional.empty()
     }
   }
 
+  override def latestSnapshotId(): Optional[raft.OffsetAndEpoch] = {
+    try {
+      Optional.of(snapshotIds.last)
+    } catch {
+      case _: NoSuchElementException =>
+        Optional.empty()
+    }
+  }
+
+  override def startSnapshotId(): Optional[raft.OffsetAndEpoch] = {
+    startSnapshotId;
+  }
+
+  override def snapshotFrozen(snapshotId: raft.OffsetAndEpoch): Unit = {
+    snapshotIds.add(snapshotId)
+  }
+
+  override def updateLogStart(logStartSnapshotId: raft.OffsetAndEpoch): Boolean = {
+    var updated = false
+    latestSnapshotId.ifPresent { snapshotId =>
+      if (startOffset < logStartSnapshotId.offset &&
+          logStartSnapshotId.offset <= snapshotId.offset &&
+          log.maybeIncrementLogStartOffset(logStartSnapshotId.offset, SnapshotGenerated)) {
+
+        log.deleteOldSegments()
+        startSnapshotId = Optional.of(logStartSnapshotId)
+        updated = true
+      }
+    }
+
+    updated
+  }
+
   override def close(): Unit = {
     log.close()
   }
 }
+
+object KafkaMetadataLog {
```

Comment:
Kind of a general note, but I think there are some logging gaps in here. I think we should tend toward the verbose side initially since snapshot events will be relatively infrequent.